In [21]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import cohen_kappa_score
from sklearn.preprocessing import StandardScaler


#Load the data and split the data into dog and cat
train=pd.read_csv('train.csv')
train_dog=train[train['Type']==1]
train_cat=train[train['Type']==2]

#All features, that do not require further preprocessing: 
#['Type', 'Age', 'Gender', 'MaturitySize', 'FurLength', 'Vaccinated', 'Dewormed', 'Sterilized', 'Health', 'Quantity', 'Fee']
feature_cols=['Type', 'Age', 'Gender', 'MaturitySize', 'FurLength', 'Vaccinated', 'Dewormed', 'Sterilized', 'Health', 'Quantity', 'Fee']

#Define y and X

y=train.AdoptionSpeed
X_unscaled=train[feature_cols]
#Scale all features
scaler=StandardScaler()
X_scaled=scaler.fit_transform(X_unscaled)
X=pd.DataFrame(X_scaled)

#Use the knn model
knn=KNeighborsClassifier(n_neighbors=54)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=42)
knn.fit(X_train, y_train)
y_pred=knn.predict(X_test)

#Print out the quadratic weighted kappa score
print(cohen_kappa_score(y_pred, y_test, weights='quadratic'))


0.2555973119706093


Scores for splitting dogs and cats, neighbors=8:
- score for both together, including feature 'Type': 0.21438
- score for only dogs: 0.19288
- score for only cats: 0.19063

Scores for other values for neighbors, dogs and cats combined:
n=48: 0.23737
n=53: 0.24947
n=55: 0.25120
n=54: 0.25560
